File: zero_knowledge/zk-test/generateProof.js
// generateProof.js
// ----------------
// Proof generator for two Noir circuits (with and without grade proof) entirely client-side.
// Exports generateProof(inputs, includeGrade) → Promise<Uint8Array> of SNARK proof bytes.

import { createFileManager, compile_program, execute_program } from '@noir-lang/noir_wasm';
import { UltraHonkBackend } from '@aztec/bb.js';

let _bbBackend = null;
async function _initBackend() {
  if (_bbBackend) return;
  _bbBackend = await UltraHonkBackend.new();
}

/**
 * Generate a ZK proof in-browser.
 * @param {Object} inputs - Circuit inputs:
 *   @property {string|number} skLo       - Lower limb of secret key
 *   @property {string|number} skHi       - Upper limb of secret key
 *   @property {string|number} leafIndex  - Leaf index in Merkle tree
 *   @property {string[]}       path       - Merkle path elements (hex strings)
 *   @property {string|number} pkX        - Public key x-coordinate
 *   @property {string|number} pkY        - Public key y-coordinate
 *   @property {string|number} collegeIdx - College index (public)
 *   @property {string|number} deptIdx    - Department index (public)
 *   @property {string|number} courseIdx  - Course index (public)
 *   @property {string|number} [grade]     - Grade (only if includeGrade=true)
 *   @property {string|number} [professor] - Professor ID (only if includeGrade=true)
 * @param {boolean} includeGrade - Selects with-grade (true) or without-grade (false) circuit
 * @returns {Promise<Uint8Array>} Raw proof bytes
 */
export async function generateProof(inputs, includeGrade) {
  await _initBackend();

  // 1) Fetch correct circuit source
  const circuitPath = includeGrade ?
    '/verifyWithGrade/main.nr' :
    '/verifyWithoutGrade/main.nr';
  const source = await fetch(circuitPath).then(r => {
    if (!r.ok) throw new Error(`Failed to load circuit: ${circuitPath}`);
    return r.text();
  });

  // 2) Compile circuit to ACIR in-memory
  const fm = createFileManager('/');
  await fm.writeFile('/main.nr', source);
  const { program } = await compile_program(fm);

  // 3) Build private inputs array
  const priv = [
    inputs.skLo,
    inputs.skHi,
    inputs.leafIndex,
    ...inputs.path,
    inputs.pkX,
    inputs.pkY
  ];

  // 4) Build public inputs array
  let pub;
  if (includeGrade) {
    if (inputs.grade == null || inputs.professor == null) {
      throw new Error('Both grade and professor are required when includeGrade=true');
    }
    pub = [
      inputs.grade,
      inputs.professor,
      inputs.collegeIdx,
      inputs.deptIdx,
      inputs.courseIdx
    ];
  } else {
    pub = [
      inputs.collegeIdx,
      inputs.deptIdx,
      inputs.courseIdx
    ];
  }

  // 5) Execute to obtain witness
  const { witness } = await execute_program(program, priv, pub);

  // 6) Generate proof bytes
  const proof = await _bbBackend.generateProof(witness);
  return proof;
}

File: zero_knowledge/circuits/verifyYesGradeYesMajor/generate_proof.py
import subprocess
import os
import argparse


def write_toml(inputs, path):
    with open(path, "w") as f:
        for key, value in inputs.items():
            if isinstance(value, list):
                # Write TOML array
                arr = ", ".join(f"\"{v}\"" for v in value)
                f.write(f"{key} = [{arr}]\n")
            else:
                # Write as a quoted string
                f.write(f"{key} = \"{value}\"\n")


def generate_proof(circuit_dir, inputs, prover_name="Prover"):
    # 1) make circuit_dir absolute
    circuit_dir = os.path.abspath(circuit_dir)

    # 2) write Prover.toml
    prover_path = os.path.join(circuit_dir, f"{prover_name}.toml")
    write_toml(inputs, prover_path)

    # 3) build the witness
    subprocess.run(
        ["nargo", "execute"],
        cwd=circuit_dir,
        check=True
    )

    # 4) locate ACIR (.json) and witness (.gz)
    target_dir   = os.path.join(circuit_dir, "target")
    acir_path    = os.path.join(target_dir, "program.json")
    witness_path = os.path.join(target_dir, "verifyWithYesGradeYesMajor.gz")

    # 5) prepare your output directory (for BB to write into)
    out_dir = os.path.join(circuit_dir, "out")
    os.makedirs(out_dir, exist_ok=True)

    # 6) run bb prove *with* scheme & oracle flags, pointing -o at the directory
    subprocess.run([
        "bb", "prove",
        "--scheme",      "ultra_honk",
        "--oracle_hash", "keccak",
        "-b",            acir_path,
        "-w",            witness_path,
        "-o",            out_dir,            # <-- this must be a directory
    ], check=True)

    # 7) now BB has created `<out_dir>/proof`
    proof_path = os.path.join(out_dir, "proof")

    # 8) read & return the raw proof bytes
    with open(proof_path, "rb") as f:
        return f.read()
    
def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Parse the input parameters required by the program."
    )

    # Simple scalar (string/integer) arguments
    parser.add_argument("--leaf_index", required=True,
                        help="Index of the Merkle-tree leaf.")
    parser.add_argument("--pk_x", required=True,
                        help="Public-key X coordinate (hex).")
    parser.add_argument("--pk_y", required=True,
                        help="Public-key Y coordinate (hex).")
    parser.add_argument("--sk_lo", required=True,
                        help="Secret-key low word (hex).")
    parser.add_argument("--sk_hi", required=True,
                        help="Secret-key high word (hex).")
    parser.add_argument("--professor", required=True, type=int,
                        help="Professor identifier.")
    parser.add_argument("--grade", required=True, type=int,
                        help="Grade value.")
    parser.add_argument("--major", required=True, type=int,
                        help="Major identifier.")
    parser.add_argument("--college_idx", required=True, type=int,
                        help="College index.")
    parser.add_argument("--dept_idx", required=True, type=int,
                        help="Department index.")
    parser.add_argument("--course_idx", required=True, type=int,
                        help="Course index.")
    parser.add_argument("--rootSchool", required=True,
                        help="Root of the Merkle tree.")

    # The Merkle proof path: variable-length positional list after --path
    parser.add_argument(
        "--path",
        nargs="+",                # one or more values
        required=True,
        metavar="HASH",
        help="Sequence of sibling hashes (hex strings) forming a Merkle proof."
    )

    return parser

if __name__ == "__main__":
    parser = build_parser()
    args = parser.parse_args()

    # Convert Namespace to ordinary dict (optional)
    inputs = vars(args)
    print("hello, generate Proof - Yes Yes is triggered")
    proof = generate_proof("../../zero_knowledge/circuits/verifyYesGradeYesMajor", inputs)
    print("∑", proof.hex())